source/common/tcp_proxy/upstream.h

```diff
@@ -196,6 +197,12 @@ class HttpUpstream : public GenericUpstream, protected Http::StreamCallbacks {
     void decodeTrailers(Http::ResponseTrailerMapPtr&& trailers) override {
```

Comment:
This feels a bit error-prone if other `decode*` callbacks can be called (e.g. `decodeMetadata` which can be out of band). Maybe it's safer to make this a caller responsibility to indicate `endStream` somehow after `decodeTrailers`, since the caller has to enforce the sequencing regardless.

Reply:
We chatted offline - if the codec always calls decodeTrailers last after decodeData(false), this is not a problem.